### PR TITLE
Make Poisson sampling opt-in in the Keras DP path

### PR DESCRIPTION
**Pull request by staticpayload**
## What this PR changes

This PR updates the Keras DP path so Poisson sampling can be done directly inside `fit()` without breaking the existing API.

The main change is an opt-in flag, `DPKerasConfig.poisson_sampling_in_fit`, which defaults to `False`. When it is enabled, the wrapper builds a Poisson-sampled `keras.utils.PyDataset` from random-access per-example arrays, validates the input sizes against the DP config, and rejects `validation_split` because the accountant needs the exact training-set size.

## Main changes

- Add `poisson_sampling_in_fit` to `DPKerasConfig`, defaulting to `False` for backward compatibility.
- Build Poisson-sampled batches internally only when that flag is enabled.
- Validate random-access array inputs and ensure `x`, `y`, and `sample_weight` agree on batch size in the Poisson-in-fit path.
- Represent padded examples through zero sample weights and derive the padding mask from `sample_weight` inside the DP training path.
- Update the Keras example and docs to show the opt-in Poisson path explicitly.
- Expand Keras tests around batch construction, padding, validation errors, and fit-argument rewriting.
- Add the empty-input padding edge case coverage in `batch_selection`.

## Why this shape

The goal here is to make the privacy-sensitive Poisson-sampling path available directly in the Keras wrapper while keeping the older workflows intact. Users who already batch or sample outside the wrapper can keep doing that. Users who want the wrapper to own batch formation can enable the flag and get a path whose sampling behavior matches the DP accounting assumptions.

## Verification

I ran the following locally in Python 3.11:

- `KERAS_BACKEND=jax python -m pytest tests/keras_api_test.py -q`
- `python -m pytest tests/batch_selection_test.py -q -k 'pad_to_multiple_of'`
- `python -m pyink --check --diff jax_privacy/keras_api.py tests/keras_api_test.py`
- `python -m pylint --rcfile=.pylintrc jax_privacy/keras_api.py`

### CODE REVIEW

**Review by ryan112358 — APPROVED** [review]

File: jax_privacy/keras_api.py [hunk 1]
```diff
@@ -77,8 +79,11 @@ class DPKerasConfig:
         noise). You should set this value before training and only based on the
         privacy guarantees you have to achieve. You should not increase the
         delta only because of poor model performance.
-      clipping_norm: The clipping norm for the per-example gradients.
-      batch_size: The batch size for the training.
+      clipping_norm: The clipping norm for the gradients.
```
Comment (dvadym) [inline jax_privacy/keras_api.py]: whe per-example was removed? clipping is performed per each example

File: jax_privacy/keras_api.py [hunk 2]
```diff
@@ -353,6 +380,218 @@ def _add_dp_sgd_attributes(model: keras.Model, params: DPKerasConfig) -> None:
...
+  leaves = jax.tree.leaves(tree)
+  if not leaves:
+    raise ValueError('Expected at least one array leaf in the training data.')
+  batch_size = None
+  for leaf in leaves:
+    if not hasattr(leaf, 'shape'):
+      raise ValueError(
+          'DP Keras training requires random-access array-like inputs.'
+      )
+    if len(leaf.shape) == 0:
```
Comment (dvadym) [inline jax_privacy/keras_api.py]: Nit (style improvement): 

```
if not leaf.shape: 

```
https://google.github.io/styleguide/pyguide.html#2144-decision

File: jax_privacy/keras_api.py [hunk 3]
```diff
@@ -353,6 +380,218 @@ def _add_dp_sgd_attributes(model: keras.Model, params: DPKerasConfig) -> None:
...
+    elif leaf_batch_size != batch_size:
+      raise ValueError(
+          'All training data leaves must agree on their leading batch'
+          ' dimension.'
+      )
+  return int(batch_size)
+
+
+def _take_batch_from_leaf(leaf: chex.Array, indices: np.ndarray) -> np.ndarray:
+  """Slices one array leaf, turning padded ``-1`` indices into zeros."""
```
Comment (dvadym) [inline jax_privacy/keras_api.py]: could you please clarify the docstring

File: jax_privacy/keras_api.py [hunk 4]
```diff
@@ -353,6 +380,218 @@ def _add_dp_sgd_attributes(model: keras.Model, params: DPKerasConfig) -> None:
...
+  if sample_weight.ndim == 2:
+    return ~jnp.any(sample_weight, axis=1)
+  raise ValueError('Expected sample_weight to be a 1D or 2D array.')
+
+
+def _maybe_symbolically_build_private_model(
+    model: keras.Model, dataset: _PoissonSampledTrainingDataset
+) -> None:
+  """Runs Keras' symbolic build before fit() sees the Poisson dataset."""
+  if not hasattr(model, '_symbolic_build'):
```
Comment (dvadym) [inline jax_privacy/keras_api.py:575]: what if we don't symbolically build?

File: jax_privacy/keras_api.py [hunk 5]
```diff
@@ -353,6 +380,218 @@ def _add_dp_sgd_attributes(model: keras.Model, params: DPKerasConfig) -> None:
...
+def _pad_batch_indices(indices: np.ndarray, multiple: int) -> np.ndarray:
+  """Pads indices with -1 so empty Poisson draws are still representable."""
+  if indices.size == 0:
+    return np.full(multiple, -1, dtype=np.int32)
+  return batch_selection.pad_to_multiple_of(indices, multiple)
+
+
+def _tree_batch_size(tree: chex.ArrayTree) -> int:
+  """Returns and validates the batch size of a pytree of arrays."""
+  leaves = jax.tree.leaves(tree)
```
Comment (dvadym) [inline jax_privacy/keras_api.py:498]: Please write down in comment expectations on `tree` (as far as I understood: all leaves have the same 0-th dimenstions

File: jax_privacy/keras_api.py [hunk 6]
```diff
@@ -671,11 +989,16 @@ def _noised_clipped_grads(
 
   noisy_grads, new_noise_state = privatizer.update(clipped_grad, noise_state)
 
-  loss = per_example_aux.values.mean()
-  unscaled_loss = per_example_aux.aux[0].mean()
+  # Open question: should these aggregates be means or sums?
```
Comment (dvadym) [inline jax_privacy/keras_api.py]: mean

File: jax_privacy/keras_api.py [hunk 7]
```diff
@@ -353,6 +380,218 @@ def _add_dp_sgd_attributes(model: keras.Model, params: DPKerasConfig) -> None:
...
+
+
+def _masked_mean(
+    values: chex.Array, is_padding_example: jax.Array
+) -> chex.Array:
+  """Averages only the non-padding examples, returning 0 for empty batches."""
+  values = jnp.asarray(values)
+  if values.ndim == 0:
+    return values
+  where = jnp.asarray(~is_padding_example)
```
Comment (dvadym) [inline jax_privacy/keras_api.py:594]: Nit (style improvement)
```
retturn values[~is_padding_example, :].mean(axis=0)
```

**Review by dvadym — COMMENTED** [review]
Thanks it looks cool. I've left some comments.

File: jax_privacy/keras_api.py (hunk 1 above)
Comment (chai1any) [reply jax_privacy/keras_api.py]: Good catch. I restored “per-example gradients” there. The clipping is still per example; I had simplified the wording too aggressively. My bad

File: jax_privacy/keras_api.py (hunk 2 above)
Comment (chai1any) [reply jax_privacy/keras_api.py]: Applied. I switched that check to if not leaf.shape.

File: jax_privacy/keras_api.py (hunk 3 above)
Comment (chai1any) [reply jax_privacy/keras_api.py]: It now says that the helper slices a batched leaf and zero-fills padded -1 index positions.

File: jax_privacy/keras_api.py (hunk 4 above)
Comment (chai1any) [reply jax_privacy/keras_api.py:575]: I expanded the comment there. The reason for the eager symbolic build is that otherwise a lazily-built model can hit the wrapped fit path before Keras has created the model and optimizer state from a standard batch.

File: jax_privacy/keras_api.py (hunk 5 above)
Comment (chai1any) [reply jax_privacy/keras_api.py:498]: Added. There is now an inline comment saying the helper expects a non-empty pytree of random-access arrays whose leaves all share the same leading batch dimension.

File: jax_privacy/keras_api.py (hunk 6 above)
Comment (chai1any) [reply jax_privacy/keras_api.py]: Agreed. I removed the open-question wording and made the masked-mean reduction explicit in the comments.

File: jax_privacy/keras_api.py (hunk 7 above)
Comment (chai1any) [reply jax_privacy/keras_api.py:594]: I left this one as-is on purpose. The boolean-indexing version is simpler to read, but it introduces a data-dependent shape, which is risky under JAX JIT and sharding. The current jnp.mean(..., where=...) path keeps the shape static while doing the same masked reduction.

Comment (dvadym) [reply jax_privacy/keras_api.py:594]: Goot point.

**Review by dvadym — APPROVED** [review]
Thanks!

**Comment by staticpayload** [comment]
To make this concrete: this was a force-push, not an additive commit on top of the old branch.

GitHub recorded the PR head moving from:
- `c74f7956ed1a2307117ccabc535845297e096b1c`

to:
- `944a6144bbf29fcfd394eb945dd35367f2396260`

at `2026-04-01T13:35:58Z`.

So the branch did move; it just moved via a history rewrite onto the latest `main`, which is why there is now a single replacement commit instead of an extra commit appended on top.

**Review by ryan112358 — APPROVED** [review]

**Comment by ryan112358** [comment]
@staticpayload It seems there was a regression in keras_api_example, can you look into it?

FYI @RamSaw 

  Traceback (most recent call last):
  DP: expected train accuracy: ~96%, val accuracy: ~92%
    File "/home/runner/work/jax_privacy/jax_privacy/examples/keras_api_example.py", line 138, in <module>
  Non-DP: expected train accuracy: ~98%, val accuracy: ~98%
      app.run(main)
    File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/site-packages/absl/app.py", line 367, in run
      _run_main(main, args)
    File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/site-packages/absl/app.py", line 312, in _run_main
      sys.exit(main(argv))
               ^^^^^^^^^^
    File "/home/runner/work/jax_privacy/jax_privacy/examples/keras_api_example.py", line 129, in main
      final_accuracy > 0.85
  AssertionError: DP Accuracy 0.8287 is too low!
